# Revisão de código

Olá! Meu nome é Alan e estou feliz em revisar seu projeto hoje!

Quando eu identificar um erro pela primeira vez, apenas o apontarei. A ideia é que você tente identificá-lo e corrigi-lo por conta própria. Ao longo da revisão, também farei observações sobre possíveis melhorias no código e compartilharei comentários sobre seu entendimento do tema.

Caso encontre dificuldades para resolver algum ponto, na próxima iteração fornecerei dicas mais específicas e exemplos práticos para ajudá-la. Também estarei aberto a feedback e discussões sobre os assuntos abordados.

Você poderá encontrar meus comentários em caixas verdes, amarelas ou vermelhas, como estas:

<div class="alert alert-block alert-success"><b>Comentário: </b> <a class="tocSkip"></a>Sucesso. Tudo está correto.</div>

<div class="alert alert-block alert-warning"><b>Comentário: </b> <a class="tocSkip"></a>Observações. Algumas recomendações.</div>

<div class="alert alert-block alert-danger"><b>Comentário: </b> <a class="tocSkip"></a>O bloco requer algumas correções. O trabalho não pode ser aceito com os comentários vermelhos.</div>

Você pode me responder usando isto:

<div class="alert alert-block alert-info"><b>Resposta da pessoa estudante    </b><a class="tocSkip"></a></div>

<div class="alert alert-block alert-warning">
<b>Comentário geral do revisor (iteração 1): </b> <a class="tocSkip"></a>

Quase lá, Tiago! A lógica de <b>quase todas</b> as 12 tarefas está correta, e você domina pontos que costumam derrubar muita gente — o índice duplo <code>user[1][0]</code> nas Tarefas 8, 10, 11 e 12, o operador <code>in</code> na Tarefa 12 e a atribuição a <code>user_age_int</code> na Tarefa 4. 


Porém, na Tarefa 7 há um pequeno caso que seu código não cobre. Coloquei um comentário em vermelho com mais detalhes. Após essa correção, seu projeto estará pronto para ser aprovado!
</div>

<div class="alert alert-block alert-success">
<b>Comentário geral do revisor (iteração 2): </b> <a class="tocSkip"></a>

Excelente! Todas as tarefas estão corretas! 

Obrigado pela correção na Tarefa 7! Parece um detalhe pequeno, mas são os tipos de bugs que a gente demora a encontrar no dia a dia. Agora seu projeto está pronto para ser aprovado! 
</div>

Uma empresa de comércio eletrônico, Store 1, começou recentemente a coletar dados sobre seus clientes. O objetivo da Store 1 é entender melhor o comportamento dos clientes e tomar decisões baseadas em dados para melhorar experiência online deles.

Como parte da equipe analítica, sua primeira tarefa é avaliar a qualidade de uma amostra de dados coletados e preparar elas para análises futuras.

# Quiz

A Store 1 visa garantir a consistência na coleta de dados. Como parte desse esforço, a qualidade dos dados coletados sobre os usuários precisa ser avaliada. Foi pedido que você revise os dados coletados e proponha alterações. Abaixo, você verá dados sobre um determinado usuário. Revise os dados e identifique possíveis problemas.

In [1]:
user_id = '32415'
user_name = ' mike_reed '
user_age = 32.0
fav_categories = ['ELECTRONICS', 'SPORT', 'BOOKS']

**Opções:**

1. O tipo de dados de `user_id` deve ser alterado de string para número inteiro (integer).
    
2. A variável `user_name` contém uma string com espaçamento desnecessário e um sublinhado entre o nome e o sobrenome.
    
3. O tipo de dados de `user_age` está incorreto.
    
4. A lista `fav_categories` contém strings em letras maiúsculas. Em vez disso, devemos converter os valores da lista para letras minúsculas.

Escreva na célula Markdown abaixo o número de opções que você identificou como problemas. Se você identificou vários problemas, separe o número por vírgulas. Por exemplo, se você acha que os números 1 e 3 estão incorretos, escreva 1, 3, e explique o motivo.

**Escreva sua resposta e explique seu raciocínio:**


# Tarefa 1

Vamos implementar as mudanças que identificamos. Primeiro, queremos corrigir os problemas com a variável `user_name`. Como verificamos, ela possui espaços desnecessários e um sublinhado como separador entre o nome e o sobrenome. Seu objetivo é remover os espaços e depois substituir o sublinhado por espaço.

In [2]:

user_name = ' mike_reed '
user_name = user_name.strip() # remova os espaços na string original
user_name = user_name.replace('_',' ')# substitua o sublinhado por espaço

print(user_name)


mike reed


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Correto e elegante! O encadeamento <code>strip().replace('_', ' ')</code> em uma única linha é a forma mais limpa de resolver, e seu comentário explicando a ordem das operações está perfeito. Lembre-se apenas de executar a célula.
</div>

# Tarefa 2

Em seguida, precisamos dividir o `user_name` atualizado em duas substrings para obter uma lista que contém dois valores: a string para o nome e a string para o sobrenome.

In [3]:
user_name = 'mike reed'
name_split = user_name.split()  # divida a string user_name aqui


print(name_split) 


['mike', 'reed']


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Correto! <code>split()</code> sem argumentos é a escolha mais robusta (trata múltiplos espaços e pontas automaticamente).
</div>

# Tarefa 3

Ótimo! Agora queremos trabalhar com a variável `user_age`. Como mencionamos antes, ela possui um tipo de dados incorretos. Vamos corrigir esse problema transformando o tipo de dados e imprimindo o resultado final.

In [4]:
user_age = 32.0
user_age = int(user_age) # altere o tipo de dados da idade de um usuário

print(user_age)

32


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Correto! Conversão de <code>float</code> para <code>int</code> feita corretamente. Detalhe de estilo (PEP 8): escreva <code>int(user_age)</code> sem espaço entre o nome da função e o parêntese.
</div>

# Tarefa 4

Como sabemos, os dados nem sempre são perfeitos. Temos que considerar cenários em que o valor de `user_age` não pode ser convertido em um número inteiro. Para evitar que nosso sistema falhe, devemos tomar medidas com antecedência.

Escreva um código que tenta converter a variável `user_age` em um número inteiro e atribua o valor transformado a `user_age_int`. Se a tentativa falhar, vamos exibir uma mensagem solicitando que o usuário forneça sua idade como um valor numérico com a mensagem: `Forneça sua idade como um valor numérico.`

In [5]:
user_age = 'thirty two'  # idade como string

try:
    user_age_int = int(user_age)
except ValueError:
    print("Forneça sua idade como um valor numérico")# escreva um código que vai tentar transformar user_age em um número inteiro e, se falhar, vai imprimir a mensagem especificada

Forneça sua idade como um valor numérico


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Correto! Você fez o ponto-chave que muitos esquecem: <b>atribuir</b> o resultado a <code>user_age_int</code> dentro do <code>try</code>. Além disso, excelente uso do `ValueError` para capturar o erro esperado.
</div>

# Tarefa 5

Por fim, observe que todas as categorias de favoritos são armazenadas em letras maiúsculas. Para preencher uma nova lista chamada `fav_categories_low` com as mesmas categorias, mas em letras minúsculas, repita os valores na lista `fav_categories`, os modifique e anexe os novos valores à lista `fav_categories_low`. Como sempre, imprima o resultado final.

In [6]:
fav_categories = ['ELECTRONICS', 'SPORT', 'BOOKS']
fav_categories_low = []

for categoria in fav_categories:
    fav_categories_low.append(categoria.lower())

# não remova a instrução de impressão abaixo
print(fav_categories_low)

['electronics', 'sport', 'books']


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Perfeito! Variável no singular (<code>category</code>), <code>lower()</code> e <code>append()</code> exatamente como esperado, com código bem legível.
</div>

# Tarefa 6

Conseguimos informações adicionais sobre os hábitos de consumo de nossos usuários, incluindo o valor gasto em cada uma de suas categorias favoritas. A administração está interessada nas seguintes métricas:

- Valor total gasto pelo usuário
- Valor mínimo gasto
- Valor máximo gasto

Vamos calcular e imprimir esses valores:

In [7]:
fav_categories_low = ['electronics', 'sport', 'books']
spendings_per_category = [894, 213, 173]

total_amount = sum(spendings_per_category) # escreva seu código aqui
max_amount = max(spendings_per_category)# escreva seu código aqui
min_amount = min(spendings_per_category)# escreva seu código aqui

# não remova a instrução de impressão abaixo
print(total_amount)
print(max_amount)
print(min_amount)

1280
894
173


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Correto! <code>sum()</code>, <code>max()</code> e <code>min()</code> usados adequadamente. Detalhe de estilo (PEP 8): remova o espaço antes do parêntese (<code>sum(...)</code> em vez de <code>sum (...)</code>).
</div>

# Tarefa 7

A empresa quer oferecer descontos aos seus clientes fiéis. Clientes que fizerem compras totalizando mais de $1.500 são considerados fiéis e vão receber um desconto.

Nosso objetivo é criar um ciclo `while` que verifique o valor total gasto e pare quando ele for atingido. Para simular novas compras, a variável `new_purchase` gera um número entre 30 e 80 em cada ciclo. Isso representa a quantidade de dinheiro gasto em uma nova compra, e é o que você precisa adicionar ao total.

Assim que o valor alvo for atingido e o ciclo `while` for encerrado, o valor final será impresso.

In [8]:
from random import randint

total_amount_spent = 1280
target_amount = 1500

while total_amount_spent <=target_amount:# escreva seu código aqui
	new_purchase = randint(30, 80) # geramos um número aleatório de 30 a 80
	total_amount_spent += new_purchase # escreva seu código aqui

print(total_amount_spent)

1505


<div class="alert alert-block alert-danger">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

O laço <code>while</code> está bem estruturado e funcional. Apenas uma observação sobre um caso extremo: o enunciado fala em <b>mais de</b> $1.500. Com <code>while ... &lt; target_amount</code>, o laço poderia parar com o total em exatamente 1500 — que ainda não é "mais de" 1500. Qual operador garantiria que ele só pare após ultrapassar o alvo?
</div>

<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Perfeito! Corrigido!
</div>

# Tarefa 8

Agora temos todas as informações sobre um cliente da maneira que queremos. A administração de uma empresa nos pediu para encontrar uma maneira de resumir toda a informação sobre um usuário. Seu objetivo é criar uma string formatada que usa informações das variáveis ​​`user_id`, `user_name` e `user_age`.

Aqui está a string final que queremos criar: `Usuário 32415 chama-se mike e tem 32 anos.`

In [9]:


user_id = '32415'
user_name = ['mike', 'reed']
user_age = 32

user_info = f"Usuário {user_id} chama-se {user_name[0]} e tem {user_age} anos."

print(user_info)

Usuário 32415 chama-se mike e tem 32 anos.


Como você já deve saber, as empresas coletam e armazenam dados de uma maneira específica. A Store 1 deseja armazenar todas as informações sobre seus clientes em uma tabela.

| user_id | user_name | user_age | purchase_category | spending_per_category |
| --- | --- | --- | --- | --- |
| '32415' | 'mike', 'reed' | 32 | 'electronics', 'sport', 'books' | 894, 213, 173 |
| '31980' | 'kate', 'morgan' | 24 | 'clothes', 'shoes' | 439, 390 |

Em termos técnicos, uma tabela é simplesmente uma lista aninhada que possui uma sublista para cada usuário.

A Store 1 criou essa tabela para seus usuários. Ela está armazenada na variável `users`. Cada sublista contém o ID do usuário, nome e sobrenome, idade, categorias favoritas e o valor gasto em cada categoria.

<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Perfeito! f-string com <code>user_name[0]</code> para acessar apenas o primeiro nome dentro da lista — exatamente o que a tarefa pede. Este é o padrão-chave do projeto e você o aplicou com segurança.
</div>

# Tarefa 9

Para calcular a receita da empresa, siga estas etapas:

1. Use um ciclo `for` para iterar na lista `users`.
2. Extraia a lista de gastos de cada usuário e some os valores.
3. Atualize o valor da receita com o total de cada usuário.

Isso vai fornecer a receita total da empresa, que você vai imprimir no final.

In [10]:
users = [
	  # este é o começo da primeira sublista
    ['32415', ['mike', 'reed'], 32, ['electronics', 'sport', 'books'],
        [894, 213, 173]
    ], # este é o fim da primeira sublista

    # este é o começo da segunda sublista
    ['31980', ['kate', 'morgan'], 24, ['clothes', 'shoes'],
        [439, 390]
    ] # este é o fim da segunda sublista
]

revenue = 0

for user in users:
	spendings_list = user[4]# extraia a lista de gastos para cada usuário e some os valores
	total_spendings = sum(spendings_list) # some os gastos em todas as categorias para obter um total para um usuário específico
	revenue += total_spendings# atualize a receita

# não remova a instrução de impressão abaixo
print(revenue)

2109


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Correto! Acumulador inicializado antes do laço, extração com <code>user[4]</code>, soma e atualização com <code>+=</code>. Execute a célula para exibir a receita (esperado: 2109). Dica: você definiu <code>spendings_list</code> mas usou <code>user[4]</code> direto no <code>sum()</code> — pode usar a variável (<code>sum(spendings_list)</code>) para aproveitar o nome descritivo que criou.
</div>

# Tarefa 10

Use um ciclo for para percorrer a lista de usuários que fornecemos e imprima os nomes dos clientes com menos de 30 anos.

In [11]:
users = [
    ['32415', ['mike', 'reed'], 32, ['electronics', 'sport', 'books'],
     [894, 213, 173]],
    ['31980', ['kate', 'morgan'], 24, ['clothes', 'books'], [439,
     390]],
    ['32156', ['john', 'doe'], 37, ['electronics', 'home', 'food'],
     [459, 120, 99]],
    ['32761', ['samantha', 'smith'], 29, ['clothes', 'electronics',
     'beauty'], [299, 679, 85]],
    ['32984', ['david', 'white'], 41, ['books', 'home', 'sport'], [234,
     329, 243]],
    ['33001', ['emily', 'brown'], 26, ['beauty', 'home', 'food'], [213,
     659, 79]],
    ['33767', ['maria', 'garcia'], 33, ['clothes', 'food', 'beauty'],
     [499, 189, 63]],
    ['33912', ['jose', 'martinez'], 22, ['sport', 'electronics', 'home'
     ], [259, 549, 109]],
    ['34009', ['lisa', 'wilson'], 35, ['home', 'books', 'clothes'],
     [329, 189, 329]],
    ['34278', ['james', 'lee'], 28, ['beauty', 'clothes', 'electronics'
     ], [189, 299, 579]],
    ]

for user in users:
    idade = user [2]
    if idade <30:
        primeiro_nome = user[1][0]
        sobrenome = user [1][1]
        print(f"{primeiro_nome}{sobrenome}")# escreva seu código aqui

katemorgan
samanthasmith
emilybrown
josemartinez
jameslee


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Correto! Filtro de idade e <code>user[1][0]</code> para imprimir só o primeiro nome.
</div>

# Tarefa 11

Vamos juntar as tarefas 9 e 10 e imprimir os nomes de usuários com menos de 30 anos com gastos totais acima de 1.000 dólares.

In [12]:
users = [
    ['32415', ['mike', 'reed'], 32, ['electronics', 'sport', 'books'],
     [894, 213, 173]],
    ['31980', ['kate', 'morgan'], 24, ['clothes', 'books'], [439,
     390]],
    ['32156', ['john', 'doe'], 37, ['electronics', 'home', 'food'],
     [459, 120, 99]],
    ['32761', ['samantha', 'smith'], 29, ['clothes', 'electronics',
     'beauty'], [299, 679, 85]],
    ['32984', ['david', 'white'], 41, ['books', 'home', 'sport'], [234,
     329, 243]],
    ['33001', ['emily', 'brown'], 26, ['beauty', 'home', 'food'], [213,
     659, 79]],
    ['33767', ['maria', 'garcia'], 33, ['clothes', 'food', 'beauty'],
     [499, 189, 63]],
    ['33912', ['jose', 'martinez'], 22, ['sport', 'electronics', 'home'
     ], [259, 549, 109]],
    ['34009', ['lisa', 'wilson'], 35, ['home', 'books', 'clothes'],
     [329, 189, 329]],
    ['34278', ['james', 'lee'], 28, ['beauty', 'clothes', 'electronics'
     ], [189, 299, 579]],
    ]


for user in users:
   idade = user [2]
   gastos = user [4]
   total_gastos = sum(gastos)
   if idade <30 and total_gastos > 1000:
       nome=user[1][0]
       print(nome)# escreva seu código aqui

samantha
james


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Perfeito! Combinação das duas condições com <code>and</code>, cálculo do total com <code>sum(gastos)</code> e impressão correta com <code>user[1][0]</code>. Como pequena sugestão, evite colocar espaços entre o nome da lista e o índice que está buscando (exemplo: <code>user [2]</code> ao invés de <code>user[2]</code>).
</div>

# Tarefa 12

Agora vamos imprimir o nome e a idade de todos os usuários que compraram roupas (clothes). Imprima o nome e a idade na mesma instrução de impressão.

In [13]:
users = [
    ['32415', ['mike', 'reed'], 32, ['electronics', 'sport', 'books'],
     [894, 213, 173]],
    ['31980', ['kate', 'morgan'], 24, ['clothes', 'books'], [439,
     390]],
    ['32156', ['john', 'doe'], 37, ['electronics', 'home', 'food'],
     [459, 120, 99]],
    ['32761', ['samantha', 'smith'], 29, ['clothes', 'electronics',
     'beauty'], [299, 679, 85]],
    ['32984', ['david', 'white'], 41, ['books', 'home', 'sport'], [234,
     329, 243]],
    ['33001', ['emily', 'brown'], 26, ['beauty', 'home', 'food'], [213,
     659, 79]],
    ['33767', ['maria', 'garcia'], 33, ['clothes', 'food', 'beauty'],
     [499, 189, 63]],
    ['33912', ['jose', 'martinez'], 22, ['sport', 'electronics', 'home'
     ], [259, 549, 109]],
    ['34009', ['lisa', 'wilson'], 35, ['home', 'books', 'clothes'],
     [329, 189, 329]],
    ['34278', ['james', 'lee'], 28, ['beauty', 'clothes', 'electronics'
     ], [189, 299, 579]],
    ]
for user in users:
    if 'clothes' in user[3]:
        nome_completo = f"{user[1][0]} {user[1][1]}"
        idade = user[2]
        print(f"nome: {nome_completo}, idade: {idade}")#escreva seu código aqui

nome: kate morgan, idade: 24
nome: samantha smith, idade: 29
nome: maria garcia, idade: 33
nome: lisa wilson, idade: 35
nome: james lee, idade: 28


<div class="alert alert-block alert-success">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Excelente! Uso do operador <code>in</code> (a solução mais elegante), <code>'clothes'</code> em minúsculas (correto!) e f-string no formato próximo ao esperado (<code>kate 24</code>).
</div>

#Escreva quaisquer comentários ou ideias finais aqui